### ISSUE TITLE
UnicodeEncode error when source contains unicode characters

### ISSUE THREAD

**Issue by github-abcde** [vpelletier/pprofile#35]
After adding the `tenacity`-module (6.2.0, from pip) to a python(2.7, sadly) project, I'm getting the following error when calling `profiler.annotate(_fh)`:
```
  File "/usr/lib/python2.7/site-packages/pprofile/__init__.py", line 676, in annotate
    }, file=out)
UnicodeEncodeError: 'ascii' codec can't encode character u'\xc9' in position 71: ordinal not in range(128)
```
It would seem that `tenacity/__init__.py:3` which contains `\xc3\x89` causes this, and I suspect this is a sign of a more general issue regarding code-as-utf8.

**Comment by vpelletier** [comment]
Thanks for the report.

The issue seems rather to be when trying to re-encode the unicode object to the `out` file (note that the error is when encoding to ascii, not when decoding the source).

With `out` a utf-8 terminal:

```
$ virtualenv -p /usr/bin/python venv
...
$ ./venv/bin/pip install tenacity
...
$ ./venv/bin/pip install pprofile
...
$ ./venv/bin/pprofile -m tenacity.__init__
...
File: /home/vincent/git/pprofile/venv/lib/python2.7/site-packages/tenacity/__init__.py
File duration: 0.00379539s (1.26%)
Line #|      Hits|         Time| Time per hit|      %|Source code
------+----------+-------------+-------------+-------+-----------
     1|         0|            0|            0|  0.00%|# -*- coding: utf-8 -*-
     2|         0|            0|            0|  0.00%|# Copyright 2016-2018 Julien Danjou
     3|         0|            0|            0|  0.00%|# Copyright 2017 Elisey Zanko
     4|         0|            0|            0|  0.00%|# Copyright 2016 Étienne Bersac
...
```
So it depends on what `_fh` is. Which is not to say that raising an exception is good, just that if that file is specifically opened as ascii or binary then changing this should be a more complete solution.

For example, pprofile (as a command) handles being redirected to a file, in which case it falls back to ascii encoding, which produces less than desirable results:
```
$ ./venv/bin/pprofile -m tenacity.__init__ > pprofile_out.txt
$ cat pprofile_out.txt
...
File: /home/vincent/git/pprofile/venv/lib/python2.7/site-packages/tenacity/__init__.py
File duration: 0.00396299s (1.31%)
Line #|      Hits|         Time| Time per hit|      %|Source code
------+----------+-------------+-------------+-------+-----------
     1|         0|            0|            0|  0.00%|# -*- coding: utf-8 -*-
     2|         0|            0|            0|  0.00%|# Copyright 2016-2018 Julien Danjou
     3|         0|            0|            0|  0.00%|# Copyright 2017 Elisey Zanko
     4|         0|            0|            0|  0.00%|# Copyright 2016 ?tienne Bersac
...
```

**Comment by github-abcde** [comment]
Sorry, I should have mentioned this: `_fh` is a filehandle as returned by `open(filename, 'a')` (in my case).
I'm profiling specific sections of a daemonized multithreaded app, and in the shutdown-code of the app, I added
```
with open(pprofile_log_filename, 'a') as _fh:
    profiler.annotate(_fh)
```
which dumps all accumulated profiling information to file. 

Reading your response, I'm considering switching to using `io.open` with `encoding='utf-8'` as a workaround on my end for this issue.

**Comment by vpelletier** [comment]
> Reading your response, I'm considering switching to using `io.open` with `encoding='utf-8'` as a workaround on my end for this issue.

This should work around this issue indeed.

> I'm profiling specific sections of a daemonized multithreaded app

On a tangential topic, for larger applications (and I guess a multithreaded daemon probably has a sizeable source) I recommend using cachegrind format, and [kcachegrind](http://kcachegrind.sourceforge.net/html/Home.html)/[qcachegrind](https://sourceforge.net/projects/qcachegrindwin/) to browse it.

If the source code is not reachable by *cachegrind (either because the machine where you read profiling results is not the same as the one where they were collected, or because the code is in zipped eggs), it is possible to build an archive with both the profiling result and all involved source code. For now it is only available in the pprofile.zope module but I'm considering moving it to base pprofile level. See the beginning of [`_iterOutFiles`](https://github.com/vpelletier/pprofile/blob/1a03a803f969b996fde504f4fc573e91659df3a1/pprofile/zope.py#L407-L430) and [`asZip`](https://github.com/vpelletier/pprofile/blob/1a03a803f969b996fde504f4fc573e91659df3a1/pprofile/zope.py#L528-L547). You can then copy this zip file to another machine and extract it (preserving internal folders) and *cachegrind should be able to find all source files for annotation.

**Comment by vpelletier** [comment]
After thinking some more about this, my feeling is that when pprofile is being called from code, I should not add magic encoding wrappers in case caller did not. I think this would make usage in python3 harder. Also, this requires introspecting the file object (ex: checking the presence of an `encoding` attribute), and I would like to limit such introspection: in my experience, various interpreter flavours (jupyter notebook comes to mind) do various things to their file descriptors.

So rather than a source fix, I did a documentation fix: 6dee08bd54ffd3092344d9465fbbbb61f4006649 documents the different requirements for `callgrind` and `annotate` methods.

> I'm considering moving it to base pprofile level.

Done: 5a3e8ffacaad90d14f115fb87e0cbfef98d2d6fb .

**Comment by github-abcde** [comment]
Thanks! That will clear up any confusion and direct future users into the right direction.
